File: python/tests/test_continuous_interactions.py
"""Test binary treatment × continuous variable interactions."""
import polars as pl
import numpy as np
import pytest

from leanfe.polars_impl import leanfe_polars
from leanfe.duckdb_impl import leanfe_duckdb


@pytest.fixture
def binary_continuous_data():
    """Generate test data with binary treatment × continuous variable."""
    np.random.seed(42)
    n = 1000
    
    df = pl.DataFrame({
        'customer_id': np.repeat(np.arange(100), 10),
        'product_id': np.tile(np.arange(50), 20),
        'treatment': np.random.binomial(1, 0.5, n),
        'price': np.random.uniform(10, 100, n),  # Continuous moderator
    })
    
    # Generate outcome with interaction effect
    # True model: y = 10 + 2*treatment + 0.5*price + 1.5*treatment*price + FE + noise
    df = df.with_columns([
        (
            10.0 +
            2.0 * pl.col('treatment') +
            0.5 * pl.col('price') +
            1.5 * pl.col('treatment') * pl.col('price') +
            pl.col('customer_id').cast(pl.Float64) * 0.01 +
            pl.lit(np.random.normal(0, 1, n))
        ).alias('revenue')
    ])
    
    return df


def test_binary_continuous_interaction_polars(binary_continuous_data):
    """Test binary treatment × continuous variable interaction (Polars)."""
    # Create interaction manually
    df = binary_continuous_data.with_columns([
        (pl.col('treatment') * pl.col('price')).alias('treatment_x_price')
    ])
    
    result = leanfe_polars(
        df,
        formula="revenue ~ treatment + price + treatment_x_price | customer_id + product_id"
    )
    
    # Check all coefficients exist
    assert 'treatment' in result['coefficients']
    assert 'price' in result['coefficients']
    assert 'treatment_x_price' in result['coefficients']
    
    # Check coefficients are close to true values
    # Note: treatment main effect has higher tolerance due to correlation with interaction term
    # when price has wide range (10-100), making main effect harder to identify precisely
    assert abs(result['coefficients']['treatment'] - 2.0) < 0.7
    assert abs(result['coefficients']['price'] - 0.5) < 0.3
    assert abs(result['coefficients']['treatment_x_price'] - 1.5) < 0.3


def test_binary_continuous_interaction_duckdb(binary_continuous_data):
    """Test binary treatment × continuous variable interaction (DuckDB)."""
    # Create interaction manually
    df = binary_continuous_data.with_columns([
        (pl.col('treatment') * pl.col('price')).alias('treatment_x_price')
    ])
    
    result = leanfe_duckdb(
        df,
        formula="revenue ~ treatment + price + treatment_x_price | customer_id + product_id"
    )
    
    # Check all coefficients exist
    assert 'treatment' in result['coefficients']
    assert 'price' in result['coefficients']
    assert 'treatment_x_price' in result['coefficients']
    
    # Check coefficients are close to true values
    # Note: treatment main effect has higher tolerance due to correlation with interaction term
    assert abs(result['coefficients']['treatment'] - 2.0) < 0.7
    assert abs(result['coefficients']['price'] - 0.5) < 0.3
    assert abs(result['coefficients']['treatment_x_price'] - 1.5) < 0.3


def test_continuous_main_effect_only(binary_continuous_data):
    """Test continuous variable as main effect (no interaction)."""
    result = leanfe_polars(
        binary_continuous_data,
        formula="revenue ~ price | customer_id + product_id"
    )
    
    # Should estimate coefficient (captures both main effect and interaction)
    assert 'price' in result['coefficients']
    assert result['coefficients']['price'] > 0


def test_multiple_continuous_variables():
    """Test multiple continuous variables."""
    np.random.seed(42)
    n = 1000
    
    df = pl.DataFrame({
        'customer_id': np.repeat(np.arange(100), 10),
        'product_id': np.tile(np.arange(50), 20),
        'price': np.random.uniform(10, 100, n),
        'quantity': np.random.uniform(1, 10, n),
    })
    
    df = df.with_columns([
        (
            10.0 +
            0.5 * pl.col('price') +
            2.0 * pl.col('quantity') +
            pl.col('customer_id').cast(pl.Float64) * 0.01 +
            pl.lit(np.random.normal(0, 1, n))
        ).alias('revenue')
    ])
    
    result = leanfe_polars(
        df,
        formula="revenue ~ price + quantity | customer_id + product_id"
    )
    
    # Check both coefficients
    assert 'price' in result['coefficients']
    assert 'quantity' in result['coefficients']
    assert abs(result['coefficients']['price'] - 0.5) < 0.3
    assert abs(result['coefficients']['quantity'] - 2.0) < 0.3

File: r/tests/test_continuous_interactions.R
#!/usr/bin/env Rscript
# Test binary treatment × continuous variable interactions

library(polars)
library(duckdb)
library(DBI)
source("R/common.R")
source("R/compress.R")
source("R/polars.R")
source("R/duckdb.R")
source("R/leanfe.R")

set.seed(42)
n <- 1000

cat("Testing binary × continuous interactions...\n\n")

# Generate test data with binary treatment × continuous variable
df <- pl$DataFrame(
  customer_id = rep(0:99, each = 10),
  product_id = rep(0:49, 20),
  treatment = rbinom(n, 1, 0.5),
  price = runif(n, 10, 100)  # Continuous moderator
)

# Generate outcome with interaction effect
# True model: y = 10 + 2*treatment + 0.5*price + 1.5*treatment*price + FE + noise
df <- df$with_columns(
  revenue = pl$lit(10.0) +
    pl$lit(2.0) * pl$col("treatment") +
    pl$lit(0.5) * pl$col("price") +
    pl$lit(1.5) * pl$col("treatment") * pl$col("price") +
    pl$col("customer_id")$cast(pl$Float64) * 0.01 +
    pl$lit(rnorm(n, 0, 1))
)

# Test 1: Binary × continuous interaction (Polars)
cat("Test 1: Binary × continuous interaction (Polars)... ")

# Create interaction manually
df_interact <- df$with_columns(
  treatment_x_price = pl$col("treatment") * pl$col("price")
)

result <- suppressWarnings(leanfe_polars(
  df_interact,
  formula = "revenue ~ treatment + price + treatment_x_price | customer_id + product_id"
))

# Check all coefficients exist
stopifnot("treatment" %in% names(result$coefficients))
stopifnot("price" %in% names(result$coefficients))
stopifnot("treatment_x_price" %in% names(result$coefficients))

# Check coefficients are close to true values
# Note: R's RNG differs from Python, so coefficients may vary more
stopifnot(result$coefficients["treatment"] > 0)  # Should be positive
stopifnot(abs(result$coefficients["price"] - 0.5) < 0.5)
stopifnot(abs(result$coefficients["treatment_x_price"] - 1.5) < 0.5)
cat("PASSED\n")

# Test 2: Binary × continuous interaction (DuckDB)
cat("Test 2: Binary × continuous interaction (DuckDB)... ")

result <- suppressWarnings(leanfe_duckdb(
  df_interact,
  formula = "revenue ~ treatment + price + treatment_x_price | customer_id + product_id"
))

stopifnot("treatment" %in% names(result$coefficients))
stopifnot("price" %in% names(result$coefficients))
stopifnot("treatment_x_price" %in% names(result$coefficients))

# Note: R's RNG differs from Python, so coefficients may vary more
# Check that coefficients have correct sign and reasonable magnitude
stopifnot(result$coefficients["treatment"] > 0)  # Should be positive
stopifnot(abs(result$coefficients["price"] - 0.5) < 0.5)
stopifnot(abs(result$coefficients["treatment_x_price"] - 1.5) < 0.5)
cat("PASSED\n")

# Test 3: Continuous main effect only
cat("Test 3: Continuous main effect only... ")

result <- suppressWarnings(leanfe_polars(
  df,
  formula = "revenue ~ price | customer_id + product_id"
))

stopifnot("price" %in% names(result$coefficients))
stopifnot(result$coefficients["price"] > 0)
cat("PASSED\n")

# Test 4: Multiple continuous variables
cat("Test 4: Multiple continuous variables... ")

df_multi <- pl$DataFrame(
  customer_id = rep(0:99, each = 10),
  product_id = rep(0:49, 20),
  price = runif(n, 10, 100),
  quantity = runif(n, 1, 10)
)

df_multi <- df_multi$with_columns(
  revenue = pl$lit(10.0) +
    pl$lit(0.5) * pl$col("price") +
    pl$lit(2.0) * pl$col("quantity") +
    pl$col("customer_id")$cast(pl$Float64) * 0.01 +
    pl$lit(rnorm(n, 0, 1))
)

result <- suppressWarnings(leanfe_polars(
  df_multi,
  formula = "revenue ~ price + quantity | customer_id + product_id"
))

stopifnot("price" %in% names(result$coefficients))
stopifnot("quantity" %in% names(result$coefficients))
stopifnot(abs(result$coefficients["price"] - 0.5) < 0.3)
stopifnot(abs(result$coefficients["quantity"] - 2.0) < 0.3)
cat("PASSED\n")

cat("\n✅ All 4 tests passed!\n")
